provider_gsuite_test.go

```diff
@@ -0,0 +1,506 @@
+package jwtauth
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"sort"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	admin "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/option"
+)
+
+const (
+	googleCredentialsEnv = "GOOGLE_CREDENTIALS"
+	gsuiteAdminEmailEnv  = "GSUITE_ADMIN_EMAIL"
+)
+
+// Tests fetching groups from G Suite using the provider configuration.
+//
+// To run the tests:
+//   1. Supply credentials via environment variables as detailed in getTestCreds()
+//   2. Supply the G Suite userName and expected groups to be fetched for the user in the test table
+func TestGSuiteProvider_FetchGroups(t *testing.T) {
+	creds, adminEmail := getTestCreds(t)
+
+	type args struct {
+		userName string
+		config   *jwtConfig
+	}
+	tests := []struct {
+		name     string
+		args     args
+		expected []interface{}
+	}{
+		{
+			name: "fetch groups from gsuite with default recursion max depth 0",
+			args: args{
+				userName: "fill_in_user_before_running",
+				config: &jwtConfig{
+					ProviderConfig: map[string]interface{}{
+						"provider":                 "gsuite",
+						"gsuite_service_account":   creds,
+						"gsuite_admin_impersonate": adminEmail,
+						"fetch_groups":             true,
+					},
+				},
+			},
+			expected: []interface{}{
+				// Fill in expected groups before running
+				// Example: "group1", "group2",
+			},
+		},
+		{
+			name: "fetch groups from gsuite with recursion max depth 1",
+			args: args{
+				userName: "fill_in_user_before_running",
+				config: &jwtConfig{
+					ProviderConfig: map[string]interface{}{
+						"provider":                 "gsuite",
+						"gsuite_service_account":   creds,
+						"gsuite_admin_impersonate": adminEmail,
+						"fetch_groups":             true,
+						"groups_recurse_max_depth": 1,
+					},
+				},
+			},
+			expected: []interface{}{
+				// Fill in expected groups before running
+				// Example: "group1", "group2",
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			b, _ := getBackend(t)
+
+			// Configure the provider
+			gProvider := new(GSuiteProvider)
+			err := gProvider.Initialize(tt.args.config)
+			assert.NoError(t, err)
+
+			// Fetch groups from G Suite
+			allClaims := map[string]interface{}{
+				"sub": tt.args.userName,
+			}
+			role := &jwtRole{
+				UserClaim:   "sub",
+				GroupsClaim: "groups",
+			}
+			groupsRaw, err := gProvider.FetchGroups(b.(*jwtAuthBackend), allClaims, role)
+			assert.NoError(t, err)
+
+			// Assert that groups are as expected
+			groupsResp, ok := normalizeList(groupsRaw)
+			assert.True(t, ok)
+			sort.Slice(groupsResp, func(i, j int) bool {
+				return groupsResp[i].(string) < groupsResp[j].(string)
+			})
+			sort.Slice(tt.expected, func(i, j int) bool {
+				return tt.expected[i].(string) < tt.expected[j].(string)
+			})
+			assert.Equal(t, tt.expected, groupsResp)
```

Comment:
BTW I think `assert.ElementsMatch()` will compare slices even if they aren't ordered the same way, so sorting them beforehand wouldn't be required.

Reply:
Nice 👍 .  Using that instead of sorting in ecd50f242cc337cb299a77057107f540291f4c7b.